packages/itwinui-react/src/core/Menu/MenuItem.tsx

```diff
@@ -107,27 +106,119 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+      if (event.parentId === nodeId) {
+        dropdownMenuContext.setLastFocusedNode({ nodeId, parentId });
+        setIsSubmenuVisible(false);
+        menuItemRef.current?.focus();
+      }
+    };
+
+    /**
+     * The `FloatingTree` needs to be updated with the nodes from the new submenu before calling this function.
+     */
+    const handleArrowRightPressed = (event: TreeEvent) => {
+      if (
+        tree?.nodesRef.current.find((n) => n.parentId === event.nodeId)?.id ===
+        nodeId
+      ) {
+        menuItemRef.current?.focus();
+      }
+    };
+
+    const handleNodeFocused = (event: TreeEvent) => {
+      // Consider a node "X" with its submenu "Y".
+      // Focusing "X" should close all submenus of "Y".
+      if (parentId === event.nodeId) {
+        setIsSubmenuVisible(false);
+      }
+
+      // When a node "X" is focused, close "X"'s siblings' submenus
+      // i.e. only one submenu in each menu can be open at a time
+      if (parentId === event.parentId && nodeId !== event.nodeId) {
+        setIsSubmenuVisible(false);
+      }
+    };
+
+    tree?.events.on('arrowLeftPressed', handleArrowLeftPressed);
+    tree?.events.on('arrowRightPressed', handleArrowRightPressed);
+    tree?.events.on('nodeFocused', handleNodeFocused);
+
+    return () => {
+      tree?.events.off('arrowLeftPressed', handleArrowLeftPressed);
+      tree?.events.off('arrowRightPressed', handleArrowRightPressed);
+      tree?.events.off('nodeFocused', handleNodeFocused);
+    };
+  }, [
+    nodeId,
+    parentId,
+    tree?.events,
+    tree?.nodesRef,
+    dropdownMenuContext,
+    children,
+  ]);
+
+  const [rightArrowPressed, setRightArrowPressed] = React.useState(false);
+
+  // Needed to trigger arrowRightPressed only after the FloatingTree has been
+  // updated with the nodes from the new submenu.
+  React.useEffect(
+    () => {
+      if (
+        rightArrowPressed &&
+        subMenuItems.length > 0 &&
+        !!tree?.nodesRef.current.find((n) => n.parentId === nodeId) // Tree has been updated with the new submenu
+      ) {
+        tree.events.emit('arrowRightPressed', {
+          nodeId,
+          parentId,
+        } satisfies TreeEvent);
+
+        setRightArrowPressed(false);
+      }
+    },
+    // TODO: Try to not require this eslint disable
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [
+      rightArrowPressed, // Run this whenever rightArrowPressed changes
+      nodeId,
+      parentId,
+      subMenuItems.length,
+      tree?.events,
+      // TODO: How to handle this warning?
+      // warning  React Hook React.useEffect has an unnecessary dependency: 'tree?.nodesRef.current'. Either exclude it or remove the dependency array. Mutable values like 'tree.nodesRef.current' aren't valid dependencies because mutating them doesn't re-render the component  react-hooks/exhaustive-deps
+      tree?.nodesRef.current,
+      children,
+      tree?.nodesRef,
+    ],
+  );
```

Comment:
I removed the `focusOnSubmenu` variable used to focus the first item in the submenu when <kbd>→</kbd> was pressed. This was because it was causing focus issues after the changes from this PR. (Related: https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1557993045)

As a replacement, I used this `useEffect`. This `useEffect` emits the `arrowRightPressed` event after certain conditions are met instead of directly emitting the `arrowRightPressed` event right after <kbd>→</kbd> is pressed. This is because we need to wait for the `FloatingTree`'s context to be updated with the new `FloatingNode`s from the submenu since `handleArrowRightPressed` needs the up-to-date tree to properly focus the first item.

This `useEffect` feels a bit hacky. Also, I disabled eslint for the deps array; not sure if that's alright. So, open to suggestions/alternatives.